src/ias_proxy/src/ias_proxy_server.rs

```diff
@@ -0,0 +1,444 @@
+          sync::Mutex};
+use utils::read_file_as_string;
+
+/// type definition for response sent from web server
+type ResponseBox = Box<Future<Item=Response<Body>, Error=Error> + Send>;
+
+/// structure defining IAS proxy server
+pub struct IasProxyServer {
+    ias_proxy_ip: String,
+    ias_proxy_port: String,
+    ias_client: Box<IasClient>,
+}
+
+/// Request body from client, proxy server may deserialize the reuqest in order to get quote if
+/// request is for attestation verification report.
+#[derive(Deserialize)]
+struct IasAVRRequestBody {
+    #[serde(rename = "isvEnclaveQuote")]
+    isv_enclave_quote: String,
+    #[serde(rename = "pseManifest")]
+    pse_manifest: String,
+    nonce: u64,
+}
+
+/// ClientResponse decoded information stored in cache
+#[derive(Debug, Clone)]
+struct IasResponse {
+    body_string: String,
+    header_map: HeaderMap,
+}
+
+lazy_static! {
+    static ref sig_rl_cache: Mutex<LruCache<String, IasResponse>> =
+        Mutex::new(LruCache::new(None));
+    static ref attestation_cache: Mutex<LruCache<String, IasResponse>> =
+        Mutex::new(LruCache::new(None));
+}
+
+const SIG_RL_LINK: &str = "/attestation/sgx/v2/sigrl";
+const AVR_LINK: &str = "/attestation/sgx/v2/report";
+const IP_PORT_DELIMITER: &str = ":";
+const UNKNOWN_ERROR_STATUS_CODE: u16 = 520;
+
+impl IasProxyServer {
+    /// Create new instance of IasProxyServer
+    fn new(
+        config: IasProxyConfig
+    ) -> Self {
+        IasProxyServer {
+            ias_proxy_ip: config.get_proxy_ip(),
+            ias_proxy_port: config.get_proxy_port(),
+            // Construct new IasClient with input config parameters
+            ias_client: Box::new(
+                IasClient::new(
+                    config.get_ias_url(),
+                    Vec::from(read_file_as_string(config.get_spid_cert_file().as_str())),
+                    None,
+                )
+            ),
+        }
+    }
+
+    /// run method to start listening on the identified IP and port
+    pub fn run(
+        &self
+    ) {
+        // Start the web server on the configured URL
+        let mut path = String::new();
+        path.push_str(self.ias_proxy_ip.as_str());
+        path.push_str(IP_PORT_DELIMITER);
+        path.push_str(self.ias_proxy_port.as_str());
+        info!("Proxy server will be started as {}", path);
+
+        // Construct socket address, panics if binding fails
+        let socket_addr: SocketAddr = match SocketAddr::from_str(&path) {
+            Ok(address_bind_successful) => address_bind_successful,
+            Err(err) => panic!("Error binding the address: {}", err),
+        };
+        info!("Socket binding successful");
+
+        // ias_client's lifetime must be static for not to clone
+        let ias_client = self.ias_client.clone();
+        // TODO: Store this server instance and call shutdown
+        let new_service = move || {
+            let ias_client = ias_client.clone();
+            service_fn(move |req|
```

Comment:
I'll try my best to give all I know to reader of code. Given that I do not want to go to internals too because hyper itself is evolving and they do not have any such explanation on this, it would otherwise turn to claim we are making on hyper's functionality.